crates/wasi/src/preview2/udp.rs

```diff
@@ -0,0 +1,87 @@
+use crate::preview2::poll::Subscribe;
+use crate::preview2::with_ambient_tokio_runtime;
+use async_trait::async_trait;
+use cap_net_ext::{AddressFamily, Blocking, UdpSocketExt};
+use io_lifetimes::raw::{FromRawSocketlike, IntoRawSocketlike};
+use std::io;
+use std::sync::Arc;
+use tokio::io::Interest;
+
+/// The state of a UDP socket.
+///
+/// This represents the various states a socket can be in during the
+/// activities of binding, and connecting.
+pub(crate) enum UdpState {
+    /// The initial state for a newly-created socket.
+    Default,
+
+    /// Binding started via `start_bind`.
+    BindStarted,
+
+    /// Binding finished via `finish_bind`. The socket has an address but
+    /// is not yet listening for connections.
+    Bound,
+
+    /// An outgoing connection is started via `start_connect`.
+    Connecting,
+
+    /// An outgoing connection is ready to be established.
+    ConnectReady,
+
+    /// An outgoing connection has been established.
+    Connected,
+}
+
+/// A host UDP socket, plus associated bookkeeping.
+///
+/// The inner state is wrapped in an Arc because the same underlying socket is
+/// used for implementing the stream types.
+pub struct UdpSocket {
+    /// The part of a `UdpSocket` which is reference-counted so that we
+    /// can pass it to async tasks.
+    pub(crate) inner: Arc<tokio::net::UdpSocket>,
+
+    /// The current state in the bind/connect progression.
+    pub(crate) udp_state: UdpState,
+}
+
+#[async_trait]
+impl Subscribe for UdpSocket {
+    async fn ready(&mut self) {
+        // Some states are ready immediately.
+        match self.udp_state {
+            UdpState::BindStarted | UdpState::ConnectReady => return,
```

Comment:
I'm a bit confused about the `ConnectReady` state here because once a socket has entered this state guests can call `send` and `receive` and the state isn't updated. This means that a socket in this state is permanently ready, which seems perhaps not intended?

Reply:
Yes, indeed, I'll fix that in TCP and here